src/pools/selection-strategies.ts

```diff
@@ -0,0 +1,118 @@
+import type { JSONValue, MessageValue } from '../utility-types'
+import { isKillBehavior, KillBehaviors } from '../worker/worker-options'
+import type { AbstractPool, IWorker } from './abstract-pool'
+
+/**
+ * Result of the round robin selection function.
+ *
+ * @template Type of the chosen element.
+ */
+export interface RoundRobinSelectionResult<Element> {
+  /** The element that was chosen. */
+  chosenElement: Element
+  /** The next calculated index. */
+  nextIndex: number
+}
+
+/**
+ * Selects the next element in a round robin selection based on the given index.
+ *
+ * @template Element Type of the element.
+ * @param elements An array of elements.
+ * @param nextIndex The next calculated index.
+ * @returns The chosen element together with the next calculated index.
+ */
+export function roundRobinSelection<Element> (
+  elements: readonly Element[],
+  nextIndex: number
+): RoundRobinSelectionResult<Element> {
+  const chosenElement = elements[nextIndex]
+  nextIndex = elements.length - 1 === nextIndex ? 0 : nextIndex + 1
+  return { chosenElement, nextIndex }
+}
+
+/**
+ * Find a free worker based on number of tasks the worker has applied.
+ *
+ * If a worker was found that has `0` tasks, it is detected as free and will be returned.
+ *
+ * If no free worker was found, `null` will be returned.
+ *
+ * @param tasks A map of tasks.
+ * @returns A free worker if there was one, otherwise `null`.
+ */
+export function findFreeWorkerBasedOnTasks<Worker> (
+  tasks: Map<Worker, number>
+): Worker | null {
+  for (const [worker, numberOfTasks] of tasks) {
+    if (numberOfTasks === 0) {
+      // A worker is free, use it
+      return worker
+    }
+  }
+  return null
+}
+
+/**
+ * Dynamically choose a worker.
+ *
+ * @param tasks tasks.
+ * @param workers workers.
+ * @param max max.
+ * @param emitter emitter.
+ * @param nextWorkerIndex nextWorkerIndex.
+ * @param nextIndexCallback nextIndexCallback.
+ * @param createAndSetupWorker createAndSetupWorker.
+ * @param registerWorkerMessageListener registerWorkerMessageListener.
+ * @param sendToWorker sendToWorker.
+ * @param destroyWorker destroyWorker.
+ * @returns The chosen one.
+ */
+export function dynamicallyChooseWorker<
```

Comment:
That one shall be a class. 
The strategy design pattern use class to define strategy for registration reason: set then execute. 
With functions the call stack is really deep and have impact on performance.

Reply:
Yes I also made a similar comment on the caller side

Reply:
Interesting, but how can I (you someone of you) do that using protected or private member functions?
The current solution preserves the member function's accessors. That's why it's build so complicated.
If we just made all of these functions `public` it would be much easier, but much more dangerous because users can do everything they want (what they already can because accessors are just TS, but they shall not...)

Reply:
Functions can just be public must not exported to the end users I guess